src/state/MerkleTree.sol

```diff
@@ -5,98 +5,193 @@ import { Arrays } from "@openzeppelin-contracts/utils/Arrays.sol";
 
 import { SHA256 } from "../libs/SHA256.sol";
 
-/// @dev This is a modified version of the OpenZeppelin MerkleTree implementation.
+/// @notice A Merkle tree implementation populating a tree of variable depth from left to right
+/// and providing on-chain Merkle proofs.
+/// @dev This is a modified version of the OpenZeppelin `MerkleTree` and `MerkleProof` implementation.
 /// https://github.com/OpenZeppelin/openzeppelin-contracts/blob/v5.2.0/contracts/utils/structs/MerkleTree.sol
+/// https://github.com/OpenZeppelin/openzeppelin-contracts/blob/v5.2.0/contracts/utils/cryptography/MerkleProof.sol
 library MerkleTree {
     struct Tree {
         uint256 _nextLeafIndex;
         mapping(uint256 level => mapping(uint256 index => bytes32 node)) _nodes;
         bytes32[] _zeros;
     }
 
+    /// @notice The hash representing the empty leaf that is not expected to be part of the tree.
     /// @dev Obtained from `sha256("EMPTY_LEAF")`.
     bytes32 internal constant _EMPTY_LEAF_HASH = 0x283d1bb3a401a7e0302d0ffb9102c8fc1f4730c2715a2bfd46a9d9209d5965e0;
 
     error TreeCapacityExceeded();
     error NonExistentLeafIndex(uint256 index);
 
+    /// @notice Sets up the tree with a capacity (i.e. number of leaves) of `2**treeDepth`
+    /// and computes the initial root of the empty tree.
+    /// @param self The tree data structure.
+    /// @param treeDepth The tree depth [0, 255].
+    /// @return initialRoot The initial root of the empty tree.
     function setup(Tree storage self, uint8 treeDepth) internal returns (bytes32 initialRoot) {
         Arrays.unsafeSetLength(self._zeros, treeDepth);
 
         bytes32 currentZero = _EMPTY_LEAF_HASH;
 
         for (uint256 i = 0; i < treeDepth; ++i) {
             Arrays.unsafeAccess(self._zeros, i).value = currentZero;
-            currentZero = SHA256.commutativeHash(currentZero, currentZero);
+            currentZero = SHA256.hash(currentZero, currentZero);
         }
 
         initialRoot = currentZero;
 
         self._nextLeafIndex = 0;
     }
 
+    /// @notice Pushes a leaf to the tree.
+    /// @param self The tree data structure.
+    /// @param leaf The leaf to add.
+    /// @return index The index of the leaf.
+    /// @return newRoot The new root of the tree.
     function push(Tree storage self, bytes32 leaf) internal returns (uint256 index, bytes32 newRoot) {
```

Comment:
As discussed on Slack, adding a method to push a batch of leaves for future improvement would be more efficient. Currently, we need to compute the root for each leaf push.

Reply:
The historical root storage is handled in the `CommitmentAccumulator`. The `push` method in `MerkleTree.sol` and the `addCommitment` method in `CommittmentAccumulator` do not store the root - the only return it.
The `ProtocolAdapter` in the end decides when to store a root 
https://github.com/anoma/evm-protocol-adapter/blob/f1d32a78c93e61202f695fc5db4002876cd65212/src/ProtocolAdapter.sol#L86-L113
and currently, this happens after we've added all commitments (see line 104 and 113 in the code snippet above).

Reply:
I'm not referring to historical root storage, but new root computation. Currently, each leaf append in `push` generates a new root with 32 hash computations.  We can recompute the root for a batch of leaf appends instead to significantly reduce the hash calls.

Reply:
Right, but we need to compute the 32 hash computations anyway to store the intermediary nodes.

https://github.com/anoma/evm-protocol-adapter/blob/3f76af92b289a4135b2d6fd8928831474e70e76b/src/state/MerkleTree.sol#L67-L68

The `newRoot` is in this sense a by-product that we can just return.

You are right that we can probably save quite some gas by having a `pushBatch` function that the protocol adapter can call instead. Maybe we can do this at some point.

However, computing 32 hashes is still cheap compared to other things that we are doing.